### PR TITLE
Disable max_distributed_connections for async_socket_for_remote.

### PR DESCRIPTION
**Pull request by KochetovNicolai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Now, if `async_socket_for_remote` is enabled, use `max_distributed_connections` only as the upper limit for the number of connections.

### CODE REVIEW

**Comment by akuzm** [comment]
Can a user still want to limit the number of connections?
Regardless of that, silently ignoring a setting is not good UX, especially if it was set explicitly. Maybe ignoring only the default would be less bad.

**Comment by KochetovNicolai** [comment]
Maybe then we should use `max_distributed_connections` as upper limit only (this setting was used only to increase the number of **threads** for distributed queries).

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 1]
```diff
@@ -1832,7 +1832,10 @@ void InterpreterSelectQuery::executeFetchColumns(QueryProcessingStage::Enum proc
     if (storage && storage->isRemote())
     {
         is_remote = true;
-        max_threads_execute_query = max_streams = settings.max_distributed_connections;
+        if (!settings.async_socket_for_remote)
+            max_threads_execute_query = max_streams = settings.max_distributed_connections;
+        else
+            max_streams = std::min<size_t>(max_streams, settings.max_distributed_connections);
```
Comment (azat) [inline src/Interpreters/InterpreterSelectQuery.cpp:1838]: But what this changes?
I mean that `max_distributed_connections` will not limit number of connections even with this patch:

```sql
SELECT sleep(1)
FROM remote('127.{2,3}', system.one)
SETTINGS max_distributed_connections = 1

Query id: 01a64150-456c-4b0c-9677-ee2e447780aa

┌─sleep(1)─┐
│        0 │
└──────────┘
┌─sleep(1)─┐
│        0 │
└──────────┘

2 rows in set. Elapsed: 1.008 sec. 

SELECT sleep(1)
FROM remote('127.{2,3}', system.one)
SETTINGS max_distributed_connections = 1, max_threads = 1

Query id: e2f633d6-dc1e-4c3c-a2e5-29b7b8cce672

┌─sleep(1)─┐
│        0 │
└──────────┘
┌─sleep(1)─┐
│        0 │
└──────────┘

2 rows in set. Elapsed: 1.009 sec. 
```

And AFAICS `max_streams` is passed to the `IStorage::read`, which will be ignored in `StorageDistributed::read` eventually.

**Comment by KochetovNicolai** [comment]
@ergifyio update

**Comment by alexey-milovidov** [comment]
> Maybe then we should use max_distributed_connections as upper limit only

Yes, it is intended to work this way.

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Comment by alexey-milovidov** [comment]
Need to test for possibe performance degradation with short queries on a cluster with 360 servers.

**Comment by alexey-milovidov** [comment]
@KochetovNicolai @azat It did not succeed. We have terrible performance degradation in this PR:

clickhouse-21.7 - master
clickhouse.5 - this PR

```
milovidov@mtexample-perftest03j:~$ time prlimit -n1000000 ./clickhouse-21.7 local --async_socket_for_remote 0 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"  
360

real    0m0.681s
user    0m0.240s
sys     0m0.584s
milovidov@mtexample-perftest03j:~$ time prlimit -n1000000 ./clickhouse-21.7 local --async_socket_for_remote 0 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"
360

real    0m0.520s
user    0m0.256s
sys     0m0.560s
milovidov@mtexample-perftest03j:~$ time prlimit -n1000000 ./clickhouse-21.7 local --async_socket_for_remote 0 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"
360

real    0m0.509s
user    0m0.208s
sys     0m0.596s
milovidov@mtexample-perftest03j:~$ time prlimit -n1000000 ./clickhouse-21.7 local --async_socket_for_remote 1 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)" 
360

real    0m0.706s
user    0m0.332s
sys     0m0.584s
milovidov@mtexample-perftest03j:~$ time prlimit -n1000000 ./clickhouse-21.7 local --async_socket_for_remote 1 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"
360

real    0m0.517s
user    0m0.352s
sys     0m0.560s
milovidov@mtexample-perftest03j:~$ time prlimit -n1000000 ./clickhouse-21.7 local --async_socket_for_remote 1 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"
360

real    0m0.556s
user    0m0.292s
sys     0m0.644s
milovidov@mtexample-perftest03j:~$ time prlimit -n1000000 ./clickhouse-21.7 local --async_socket_for_remote 1 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"
360

real    0m0.539s
user    0m0.324s
sys     0m0.596s
milovidov@mtexample-perftest03j:~$ time prlimit -n1000000 ./clickhouse-21.7 local --async_socket_for_remote 1 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"
360

real    0m0.531s
user    0m0.224s
sys     0m0.700s
milovidov@mtexample-perftest03j:~$ time prlimit -n1000000 ./clickhouse.5 local --async_socket_for_remote 0 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"
360

real    0m0.776s
user    0m0.304s
sys     0m0.496s
milovidov@mtexample-perftest03j:~$ time prlimit -n1000000 ./clickhouse.5 local --async_socket_for_remote 0 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"
360

real    0m0.487s
user    0m0.236s
sys     0m0.508s
milovidov@mtexample-perftest03j:~$ time prlimit -n1000000 ./clickhouse.5 local --async_socket_for_remote 0 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"
360

real    0m0.496s
user    0m0.348s
sys     0m0.468s
milovidov@mtexample-perftest03j:~$ time prlimit -n1000000 ./clickhouse.5 local --async_socket_for_remote 1 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"
360

real    0m1.296s
user    0m0.164s
sys     0m0.276s
milovidov@mtexample-perftest03j:~$ time prlimit -n1000000 ./clickhouse.5 local --async_socket_for_remote 1 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"
360

real    0m1.305s
user    0m0.116s
sys     0m0.320s
milovidov@mtexample-perftest03j:~$ time prlimit -n1000000 ./clickhouse.5 local --async_socket_for_remote 1 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"
360

real    0m1.299s
user    0m0.200s
sys     0m0.236s
```

**Comment by alexey-milovidov** [comment]
@Avogar Need to investigate, why async_socket_for_remote did not help.

**Comment by alexey-milovidov** [comment]
Maybe it is due to DNS requests and it will be ok for clickhouse-server due to DNS caching?

**Comment by alexey-milovidov** [comment]
It is slow:

```
mtexample-perftest03j.yandex.ru :) SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

SELECT count()
FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

Query id: 426fbfdf-02e2-42b1-b0aa-a655752ee1eb

┌─count()─┐
│     360 │
└─────────┘

1 rows in set. Elapsed: 1.372 sec. 

mtexample-perftest03j.yandex.ru :) SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

SELECT count()
FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

Query id: e7639dae-e14d-4a3f-ad83-c84c85f75505

┌─count()─┐
│     360 │
└─────────┘

1 rows in set. Elapsed: 1.197 sec. 

mtexample-perftest03j.yandex.ru :) SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

SELECT count()
FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

Query id: 56ba6049-18ec-429d-8941-b29067f315ed

┌─count()─┐
│     360 │
└─────────┘

1 rows in set. Elapsed: 1.289 sec. 

mtexample-perftest03j.yandex.ru :) SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

SELECT count()
FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

Query id: 1dd37d7f-af3f-4e91-9293-c6a599e1472b

┌─count()─┐
│     360 │
└─────────┘

1 rows in set. Elapsed: 1.196 sec. 

mtexample-perftest03j.yandex.ru :) SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

SELECT count()
FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

Query id: 4bc30a3d-d2a1-49f9-833f-471d01b7dd38

┌─count()─┐
│     360 │
└─────────┘

1 rows in set. Elapsed: 1.198 sec. 

mtexample-perftest03j.yandex.ru :) SET async
async_socket_for_remote  asynchronous_metrics
mtexample-perftest03j.yandex.ru :) SET async_socket_for_remote = 0

SET async_socket_for_remote = 0

Query id: 7dd4e931-075e-498e-bc32-dc0c780bacc2

Ok.

0 rows in set. Elapsed: 0.001 sec. 

mtexample-perftest03j.yandex.ru :) SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

SELECT count()
FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

Query id: 75d1ba22-1e24-46bb-8e54-79194114e4ad

┌─count()─┐
│     360 │
└─────────┘

1 rows in set. Elapsed: 0.350 sec. 

mtexample-perftest03j.yandex.ru :) SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

SELECT count()
FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

Query id: 556f44ab-e6b1-4efe-b28c-984a2ef6bddf

┌─count()─┐
│     360 │
└─────────┘

1 rows in set. Elapsed: 0.331 sec. 

mtexample-perftest03j.yandex.ru :) SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

SELECT count()
FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)

Query id: 4a16c9da-d647-4a79-8404-f80595bdf0f2

┌─count()─┐
│     360 │
└─────────┘

1 rows in set. Elapsed: 0.327 sec. 

mtexample-perftest03j.yandex.ru :)
```

**Comment by azat** [comment]
>@KochetovNicolai @azat It did not succeed. We have terrible performance degradation in this PR:

I guess this is because the connect is slow, and after this patch connect will be done from not more then max_threads threads (`max_threads_execute_query`)
This is not a problem for regular Distributed tables, since they have a connection pool and should not have lots of connects.

**Comment by azat** [comment]
>Now, if async_socket_for_remote is enabled, use max_distributed_connections only as the upper limit for the number of connections.

But I'm still have some misunderstanding I guess, AFAICS max_distributed_connections does not limit number of connects, so they will be established anyway, and so if user think that using async_socket_for_remote=1 and max_distributed_connections=1 will go through list of shards one by one (suppose user have lots of such queries and he don't want to send them at once to all shards) this will not happen (see detalis).

<details>

```sql
:) SELECT sleep(1)　FROM remote('127.{2,3}', system.one)　SETTINGS max_distributed_connections = 1, max_threads=1, async_socket_for_remote=1

SELECT sleep(1)
FROM remote('127.{2,3}', system.one)
SETTINGS max_distributed_connections = 1, max_threads = 1, async_socket_for_remote = 1

Query id: fc142d5a-d51e-4147-8373-ceaae1ff8229

┌─sleep(1)─┐
│        0 │
└──────────┘
┌─sleep(1)─┐
│        0 │
└──────────┘

2 rows in set. Elapsed: 1.051 sec.

:) SELECT sleep(1)　FROM remote('127.{2,3}', system.one)　SETTINGS max_distributed_connections = 1, max_threads=1, async_socket_for_remote=0

SELECT sleep(1)
FROM remote('127.{2,3}', system.one)
SETTINGS max_distributed_connections = 1, max_threads = 1, async_socket_for_remote = 0

Query id: 0f692605-cd68-4ca5-a324-150278218ac4

┌─sleep(1)─┐
│        0 │
└──────────┘
┌─sleep(1)─┐
│        0 │
└──────────┘

2 rows in set. Elapsed: 2.067 sec.

```

</details>

**Comment by KochetovNicolai** [comment]
@alexey-milovidov it is not surprising that your query works slower. With new changes, we use less threads for async connections (16 in your example). We need some time to send 360 connections using 16 threads:
```
$ time prlimit -n1000000 ./21.8.1.7401/usr/bin/clickhouse local --async_socket_for_remote 1 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"  
360

real	0m1.373s
user	0m0.136s
sys	0m0.324s

$ time prlimit -n1000000 ./21.8.1.7401/usr/bin/clickhouse local --max_threads=32 --async_socket_for_remote 1 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"  
360

real	0m0.963s
user	0m0.172s
sys	0m0.320s

$ time prlimit -n1000000 ./21.8.1.7401/usr/bin/clickhouse local --max_threads=64 --async_socket_for_remote 1 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"  
360

real	0m0.708s
user	0m0.176s
sys	0m0.360s

$ time prlimit -n1000000 ./21.8.1.7401/usr/bin/clickhouse local --max_threads=1024 --async_socket_for_remote 1 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"  
360

real	0m0.673s
user	0m0.592s
sys	0m0.988s
```

the sane result if we limit the number of threads without async:
```
$ time prlimit -n1000000 ./21.8.1.7401/usr/bin/clickhouse local --max_threads=16 --max_distributed_connections=16 --async_socket_for_remote 0 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"  
360

real	0m1.617s
user	0m0.152s
sys	0m0.312s
$ time prlimit -n1000000 ./21.8.1.7401/usr/bin/clickhouse local --max_threads=32 --max_distributed_connections=32 --async_socket_for_remote 0 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"  
360

real	0m1.023s
user	0m0.144s
sys	0m0.328s
$ time prlimit -n1000000 ./21.8.1.7401/usr/bin/clickhouse local --max_threads=64 --max_distributed_connections=64 --async_socket_for_remote 0 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"  
360

real	0m0.966s
user	0m0.192s
sys	0m0.316s
$ time prlimit -n1000000 ./21.8.1.7401/usr/bin/clickhouse local --max_threads=1024 --max_distributed_connections=1024 --async_socket_for_remote 0 --query "SELECT count() FROM remote('mtexample{001..120}-{1..3}.metrika.yandex.net', system.one)"  
360

real	0m0.671s
user	0m0.448s
sys	0m0.840s
```

**Comment by alexey-milovidov** [comment]
@KochetovNicolai The purpose of `async_socket_for_remote` was to perform multiple connections in parallel from every single thread.

**Comment by alexey-milovidov** [comment]
If regular Distributed tables did not slow down, we can merge this PR.

**Comment by alexey-milovidov** [comment]
> We need some time to send 360 connections using 16 threads

360 connections can be performed concurrently even from single thread. The time required should be the same as for slowest server.

**Pull request closed without merging**